### PR TITLE
Resolve components by module ID during compilation

### PR DESCRIPTION
**Pull request by tony-sull**
## Changes

This updates Astro to resolve `.astro` components by Vite's module ID, fixing an issue using `.astro` and framework components in shared NPM packages

**Details**

During the compilation step, Astro was parsing the file path for a component and expecting astro components to always live in the project's root directory.

This caused a few weird corner cases for NPM libraries that use both `.astro` and framework components internally, ex: a theme that is mostly `.astro` with a few internal react components for interactivity.

## Testing

Added a test suite that uses components from a separate package in the monorepo.  Without the fix this test suite fails to build when the component files can't be found

## Docs

None, but I will open a separate PR to update our `component` starter once this fix is published

### CODE REVIEW

File: packages/astro/src/core/build/internal.ts [hunk 1]
```diff
@@ -80,17 +79,15 @@ export function trackPageData(
...
-	clientOnlys: string[],
-	astroConfig: AstroConfig
+	clientOnlys: string[]
 ) {
 	for (const clientOnlyComponent of clientOnlys) {
-		const coPath = viteID(new URL('.' + clientOnlyComponent, astroConfig.root));
 		let pageDataSet: Set<PageBuildData>;
-		if (internals.pagesByClientOnly.has(coPath)) {
-			pageDataSet = internals.pagesByClientOnly.get(coPath)!;
+		if (internals.pagesByClientOnly.has(clientOnlyComponent)) {
```
Comment (tony-sull) [inline packages/astro/src/core/build/internal.ts:86]: `transform()` is now using the module ID directly rather than the project root relative path, no need to convert to here with `viteID`

File: packages/astro/src/core/build/internal.ts [hunk 2]
```diff
@@ -115,8 +112,9 @@ export function* getPageDatasByClientOnlyChunk(
 	const pagesByClientOnly = internals.pagesByClientOnly;
 	if (pagesByClientOnly.size) {
 		for (const [modulePath] of Object.entries(chunk.modules)) {
-			if (pagesByClientOnly.has(modulePath)) {
-				for (const pageData of pagesByClientOnly.get(modulePath)!) {
+			const pathname = `/@fs${prependForwardSlash(modulePath)}`;
+			if (pagesByClientOnly.has(pathname)) {
```
Comment (tony-sull) [inline packages/astro/src/core/build/internal.ts:116]: module path is now the full module ID, prepending `/@fs` to match the use in `transform()` to match the `trackClientOnlyPageDatas()` use (above)

File: packages/astro/src/core/build/vite-plugin-hoisted-scripts.ts [hunk 3]
```diff
@@ -25,7 +25,11 @@ export function vitePluginHoistedScripts(
 			if (virtualHoistedEntry(id)) {
 				let code = '';
 				for (let path of internals.hoistedScriptIdToHoistedMap.get(id)!) {
-					code += `import "${path}";`;
+					let importPath = path;
+					if (importPath.startsWith('/@fs')) {
+						importPath = importPath.slice('/@fs'.length);
+					}
+					code += `import "${importPath}";`;
```
Comment (tony-sull) [inline packages/astro/src/core/build/vite-plugin-hoisted-scripts.ts:32]: strip off the `/@fs` prefix before building the hoisted script's `import` statement

File: packages/astro/src/vite-plugin-astro/compile.ts [hunk 4]
```diff
@@ -34,16 +34,25 @@ function safelyReplaceImportPlaceholder(code: string) {
...
+interface CompileProps {
+	config: AstroConfig;
+	filename: string;
+	moduleId: string;
+	source: string;
+	ssr: boolean;
+	viteTransform: TransformHook;
+}
+
+async function compile({
```
Comment (tony-sull) [inline packages/astro/src/vite-plugin-astro/compile.ts:46]: (opinionated) using a single object prop to clean up the method props

**Comment by aFuzzyBear** [comment]
Thank you @tony-sull  really appreciate this fix dude

**Comment by delucis** [comment]
Dang, was really hopeful this might also solve #3053, but it doesn’t look like it does. Great work nevertheless!

File: packages/astro/test/fixtures/component-library/package.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "@test/component-library",
+  "version": "0.0.0",
+  "private": true,
+  "dependencies": {
+    "astro": "workspace:*",
+    "@astrojs/preact": "workspace:*",
+    "@astrojs/react": "workspace:*",
+    "@astrojs/svelte": "workspace:*",
+    "@test/component-library-shared": "workspace:*",
```
Comment (bholmesdev) [inline packages/astro/test/fixtures/component-library/package.json:10]: I'm always a little weary of `workspace:*` as a package test. Is it possible to generate a tarball from `component-library-shared` via `pnpm pack`, and reference that file from here? This would 100% mock an external npm dependency

**Comment by bholmesdev** [comment]
Also, does this resolve any reported bugs? Didn't see any in the description so double checking!

File: packages/astro/test/fixtures/component-library/package.json (hunk 5 above)
Comment (tony-sull) [reply packages/astro/test/fixtures/component-library/package.json:10]: I'm not actually sure if tarballs work properly in vite 🤔 I don't use `pnpm pack` often so I may have it misconfigured, but I tried this locally and Vite blew up with errors failing to resolve anything inside the tarball dependency

`link:../component-library-shared` would work for the tests, but I'm not sure the details of how that's different from `workspace:*`

**Comment by tony-sull** [comment]
@bholmesdev I still need to make a pass through all the open issues to see if any seem related. I can't help but think this was the underlying cause of a few strange bugs, but I ran into this when working on the docs template and forgot to file my own issue to track before diving into the fix

**Comment by tony-sull** [comment]
> Dang, was really hopeful this might also solve #3053, but it doesn’t look like it does. Great work nevertheless!

I was hoping the same 🙃 Hopefully this gets us closer though, I'd really like to see that issue fixed soon as well!

**Review by matthewp — APPROVED** [review]
